### ISSUE TITLE
Why do not mask padding?

### ISSUE THREAD

**Issue by PanYZhu** [pmixer/SASRec.pytorch#27]
In model.py, I'm confused about setting the padding part to zero instead of -inf.
https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L72
https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L73
why not pass timeline_mask to torch.nn.MultiheadAttention as key_padding_mask?
https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L83

**Comment by pmixer** [comment]
@PanYZhu I'm not very clear about the question temporally, maybe we can start from syncing information about padding, there're two kinds of needs for paddings:

1. disable placeholder item from effecting final result
2. enforce causality in sequence processing, only previous items can effect next item prediction result

and commonly we just pad 0, like use 0th item as placeholder item, -inf is for minimize attention weights in MHA which is a speical case.

so now, could u pls elaborate more into `setting the padding part to zero instead of -inf.`? does that means you suggest pad using -inf instead of 0 in model.py line 72 and 73?

for `why not pass timeline_mask to torch.nn.MultiheadAttention as key_padding_mask?`, I need sometime to recall why I commented it.

thx for taking your time on reading the code!

**Comment by GLA1VE11** [comment]
I also think there is a problem with mask here, if it only becomes 0, even if there is attn_mask, the **word** behind will see the **padding** in front, and the part of padding will be calculated into attention.

**Comment by GLA1VE11** [comment]
So I think we should use both key_padding_mask and attn_mask, and then replace those 'nan' with 0（because if we use both of them, the padding will becomes 'nan'）

**Comment by GLA1VE11** [comment]
> In model.py, I'm confused about setting the padding part to zero instead of -inf.
> 
> https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L72
> 
> 
> https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L73
> 
> 
> why not pass timeline_mask to torch.nn.MultiheadAttention as key_padding_mask?
> https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L83

I think that because the author use 'attn_mask', then if we use 'key_padding_mask', chances are that there will be a row that are all masked(like the first row if it is a padding), and then after softmax, the attention weight will be nan.

> In model.py, I'm confused about setting the padding part to zero instead of -inf.
> 
> https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L72
> 
> 
> https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L73
> 
> 
> why not pass timeline_mask to torch.nn.MultiheadAttention as key_padding_mask?
> https://github.com/pmixer/SASRec.pytorch/blob/4297d0950c6a6bffba92ecd4d1bf7204d07eb8c7/model.py#L83


I think that because the author use 'attn_mask', then if we use 'key_padding_mask', chances are that there will be a row that are all masked(like the first row if it is a padding), and then after softmax, the attention weight will be nan.

**Comment by PanYZhu** [comment]
@pmixer thx for your reply! I appreciate for your patient explanation. I misunderstood some knowledge about padding before and I found there was no wrong with the padding part now.

**Comment by GLA1VE11** [comment]
> @pmixer thx for your reply! I appreciate for your patient explanation. I misunderstood some knowledge about padding before and I found there was no wrong with the padding part now.

Could you please explain why not use the key_padding_mask?

**Comment by pmixer** [comment]
> > @pmixer thx for your reply! I appreciate for your patient explanation. I misunderstood some knowledge about padding before and I found there was no wrong with the padding part now.
> 
> Could you please explain why not use the key_padding_mask?

@GLA1VE11 I could hardly recall the concrete details as the code was written more than 2 years ago.

If u mean:

https://github.com/pmixer/SASRec.pytorch/blob/d192a40203ea1391410f71445b863fc3cc3e2620/model.py#L83

https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html mentioned that at most one of key_padding_mask or attn_mask is passed, I choose to use attn_mask for consistency with original tf version, not completely sure if these two mask params could used as alternatives for the case, try it if you are interested.

Thx.
